airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/AbstractDestination.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.base;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.SyncMode;
+import io.airbyte.queue.BigQueue;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class AbstractDestination implements Destination, DestinationWriteOperations {
```

Comment:
I think it's telling that BQ isn't using this abstraction. The goal in my mind for the AbstractDestination was that it was supposed to contain the general logic that we do the same for _every_ destination. If BQ isn't using it, it means we haven't really found much to abstract at this point. Looking at this class now, I'm not sure how as a developer one would approach trying to leverage it. From what I am seeing it seems like:
1. so either it should change so that it is useable by BQ
1. or it doesn't really provide much value now, we should get rid of it and just keep the `AbstractJdbcDestination`

Going to tinker with this a little bit now before finishing the review.

Reply:
I played with this some more. I think there might be some other abstraction we can find here, but it doesn't seem super high value. I think it'll be come clearer with time. For now, I think I'd recommend getting rid of the `AbstractDestination` abstraction and collapsing it into the `AbstractJdbcDestination` one. On its own right now I don't think `AbstractJdbcDestination` is adding value. On top of that it adds a lot of complexity. What do you think?

Reply:
To add a little more detail to what I mean about it adding complexity, as a developer it is really hard for me to figure out what parts of `AbstractJdbcDestination` I am supposed to override. It has a lot of methods that are potentially overridable. Very few of them are required at compile time, so it's hard to figure out how I am supposed to use it.

Reply:
put some time on the calendar to follow up on this in the morning SF time.

Reply:
I didnt touch bigquery/csv destinations on purposes but sure we could refactor them too